client/packages/purchase_orders/src/DetailView/ImportLines/UploadTab.tsx

```diff
@@ -0,0 +1,337 @@
+import { ImportPanel } from './ImportPanel';
+import * as PurchaseOrderLineImportModal from './PurchaseOrderLineImportModal';
+import { importPurchaseOrderLinesToCsv } from '../utils';
+
+interface UploadTabProps {
+  setEquipment: React.Dispatch<React.SetStateAction<ImportRow[]>>;
+  setErrorMessage: (value: React.SetStateAction<string>) => void;
+  setWarningMessage: (value: React.SetStateAction<string>) => void;
+  onUploadComplete: () => void;
+}
+
+// introduce new interface to accommodate dynamic keys of parsed result
+interface ParsedAsset {
+  id: string;
+  [key: string]: string | undefined;
+}
+
+function getImportHelpers<T, P>(
+  row: P,
+  rows: T[],
+  index: number,
+  t: TypedTFunction<LocaleKey>
+) {
+  const importRow = {
+    id: FnUtils.generateUUID(),
+    properties: {},
+  } as T;
+  const rowErrors: string[] = [];
+  const rowWarnings: string[] = [];
+
+  const addCell = (
+    key: keyof T,
+    localeKey: LocaleKey,
+    formatter?: (value: string) => unknown
+  ) => {
+    const prop = t(localeKey) as keyof P;
+    const value = row[prop] ?? '';
+    if (value !== undefined) {
+      (importRow[key] as unknown) = formatter
+        ? formatter(value as string)
+        : value;
+    }
+  };
+
+  const addRequired = (
+    key: keyof T,
+    localeKey: LocaleKey,
+    formatter?: (value: string) => unknown
+  ) => {
+    const prop = t(localeKey) as keyof P;
+    const value = row[prop] ?? '';
+
+    if (value === undefined || (value as string).trim() === '') {
+      rowErrors.push(
+        t('error.field-must-be-specified', {
+          field: t(localeKey),
+        })
+      );
+      return;
+    }
+
+    addCell(key, localeKey, formatter);
+  };
+
+  const addSoftRequired = (
+    key: keyof T,
+    localeKey: LocaleKey,
+    formatter?: (value: string) => unknown
+  ) => {
+    const prop = t(localeKey) as keyof P;
+    const value = row[prop] ?? '';
+
+    if (value === undefined || (value as string).trim() === '') {
+      rowWarnings.push(
+        t('warning.field-not-parsed', {
+          field: t(localeKey),
+        })
+      );
+      return;
+    }
+
+    if (
+      formatter &&
+      value &&
+      (formatter(value as string) === undefined ||
+        formatter(value as string) === null)
+    ) {
+      rowWarnings.push(
+        t('warning.field-not-parsed', {
+          field: t(localeKey),
+        })
+      );
+      return;
+    }
+    addCell(key, localeKey, formatter);
+  };
+
+  const addUnique = (
+    key: keyof T,
+    localeKey: LocaleKey,
+    formatter?: (value: string) => unknown
+  ) => {
+    const prop = t(localeKey) as keyof P;
+    const value = row[prop] ?? '';
+
+    addRequired(key, localeKey, formatter);
+
+    // check for duplicates
+    if (rows.some((r, i) => r[key] === value && index !== i)) {
+      rowErrors.push(
+        t('error.duplicated-field', {
+          field: t(localeKey),
+        })
+      );
+    }
+  };
+
+  const addUniqueCombination = (
```

Comment:
New validation checker which checks for unique combination.

I thought could be helpful to refactor this with hash map which could be quicker if uploading many lines, but seems to work OK for now